icu4c/source/common/unicode/locid.h

```diff
@@ -1217,12 +1224,59 @@ class U_COMMON_API_CLASS Locale : public UObject {
     static_assert(sizeof(Nest) == Nest::SIZE);
 
     /**
-     * Locale data that needs to be heap allocated in a std::unique_ptr object.
+     * Locale data that needs to be heap allocated in the union Payload object.
      * @internal
      */
     struct Heap;
 
-    std::variant<std::monostate, Nest, std::unique_ptr<Heap>> payload;
+    /**
+     * This is kind of std::variant but customized to not waste any space on the
+     * discriminator or on any padding, and to copy any heap allocated object.
+     * @internal
+     */
+    union Payload {
+      private:
+        Nest nest;
+
+        struct HeapPtr {
+            ELocaleType type;
+            Heap* ptr;
+
+            HeapPtr& operator=(const Heap& other);
+            HeapPtr& operator=(HeapPtr&& other) noexcept;
+        } heap;
```

Comment:
I find it confusing to read the impl code where `heap` is not of type `Heap`. Please rename this to heapPtr.